pkg/npm/npm.go

```diff
@@ -9,58 +9,128 @@ import (
-	// filename, since it's meant to be uploaded directly to npm, which means that we have to get
-	// that information by parsing the output of the command.
+	// Note that `npm pack` doesn't have the ability to specify the resulting filename, since
+	// it's meant to be uploaded directly to npm, which means we have to get that information
+	// by parsing the output of the command. However, if we're using `yarn`, we can specify a
+	// filename.
+	var packfile string
+	if !npm {
+		packfile = fmt.Sprintf("%s.tgz", uuid.NewV4().String())
+		c.Args = append(c.Args, "--filename", packfile)
+	}
+
+	// Run the command.
+	// `stdout` is ignored for the command, as it does not generate useful data.
 	var stdout bytes.Buffer
-	c.Stdout = &stdout
-	c.Stderr = os.Stderr
-	if err := c.Run(); err != nil {
+	if err = runCmd(c, npm, &stdout, stderr); err != nil {
 		return nil, err
 	}
-	packfile := strings.TrimSpace(stdout.String())
+
+	// If `npm` was used, parse the filename from the output.
+	if npm {
+		packfile = strings.TrimSpace(stdout.String())
+	}
+
 	defer os.Remove(packfile)
 
 	packTarball, err := ioutil.ReadFile(packfile)
 	if err != nil {
-		return nil, fmt.Errorf(
-			"npm pack completed successfully but the packed .tar.gz file was not generated")
+		return nil, fmt.Errorf("%s pack completed successfully but the packed .tgz file was not generated", bin)
 	}
 
 	return packTarball, nil
 }
 
 // Install runs `npm install` in the given directory, installing the dependencies for the Node.js
-// app located there.
-func Install(dir string, stdout, stderr io.Writer) error {
-	// We pass `--loglevel=error` to prevent `npm` from printing warnings about missing
-	// `description`, `repository`, and `license` fields in the package.json file.
-	c := exec.Command("npm", "install", "--loglevel=error")
+// app located there. If the `PULUMI_PREFER_YARN` environment variable is set, `yarn install` is used
+// instead of `npm install`.
+func Install(dir string, stdout, stderr io.Writer) (string, error) {
+	c, npm, bin, err := getCmd("install")
+	if err != nil {
+		return bin, err
+	}
 	c.Dir = dir
 
 	// Run the command.
-	c.Stdout = stdout
-	c.Stderr = stderr
-	if err := c.Run(); err != nil {
-		return err
+	if err = runCmd(c, npm, stdout, stderr); err != nil {
+		return bin, err
 	}
 
 	// Ensure the "node_modules" directory exists.
 	if _, err := os.Stat("node_modules"); os.IsNotExist(err) {
-		return errors.Errorf("npm install reported success, but node_modules directory is missing")
+		return bin, errors.Errorf("%s install reported success, but node_modules directory is missing", bin)
+	}
+
+	return bin, nil
+}
+
+// getCmd returns the `npm` or `yarn` command depending on whether `PULUMI_PREFER_YARN` is set and yarn can be found
+// on the $PATH. If `npm` is being used, true is returned (false for `yarn`), and the file (either "npm" or "yarn").
+func getCmd(command string) (*exec.Cmd, bool, string, error) {
```

Comment:
The comment above this file is super confusing. I think we can simplify the explanation to just:

// getCmd returns the exec.Cmd used to install NPM dependencies. It will either use npm or yarn depending
// on what is available on the current path, and if PULUMI_PREFER_YARN is truth.
// The boolean return parameter indicates if npm is chosen or not (instead of yarn).

Reply:
Fixed.